### PR TITLE
[SPARK-4346][SPARK-3596][YARN] Commonize the monitor logic

### PR DESCRIPTION
**Pull request by Sephiroth-Lin**
1. YarnClientSchedulerBack.asyncMonitorApplication use Client.monitorApplication so that commonize the monitor logic
2. Support changing the yarn client monitor interval, see #5292
3. More details see discussion on https://github.com/apache/spark/pull/3143

### CODE REVIEW

**Comment by srowen** [comment]
Jenkins, add to whitelist

**Comment by srowen** [comment]
ok to test

**Comment by Sephiroth-Lin** [comment]
Jenkins, retest please

**Comment by srowen** [comment]
CC a few people who have touched this bit of the code: @kasjain @witgo @andrewor14

File: yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala [hunk 1]
```diff
@@ -127,23 +127,11 @@ private[spark] class YarnClientSchedulerBackend(
...
-          }
-          if (state == YarnApplicationState.FINISHED ||
-            state == YarnApplicationState.KILLED ||
-            state == YarnApplicationState.FAILED) {
-            logError(s"Yarn application has already exited with state $state!")
-            sc.stop()
-            stopping = true
-          }
-          Thread.sleep(1000L)
+        val (state, _) = client.monitorApplication(appId, logApplicationReport = false)
```
Comment (witgo) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: The process doesn't look the same. `ApplicationNotFoundException` is not handled.

Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: Yes, we can just add it in `Client#monitorApplication` as well

**Comment by andrewor14** [comment]
Hi @Sephiroth-Lin this looks good for the most part. The only reason why I didn't merge these code paths initially is because the loop here checks on `stopping`, and I didn't want to pass the boolean into `monitorApplication`. However, on second examination this is a daemon thread anyway, so it will eventually exit regardless of whether the backend has been stopped.

Once you address @witgo's comment I will merge this thanks.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 2]
```diff
@@ -559,50 +560,56 @@ private[spark] class Client(
...
-          .mkString("")
-
-        // If DEBUG is enabled, log report details every iteration
-        // Otherwise, log them every time the application changes state
-        if (log.isDebugEnabled) {
-          logDebug(formattedDetails)
-        } else if (lastState != state) {
-          logInfo(formattedDetails)
+      try {
+        val report = getApplicationReport(appId)
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: you probably don't need to wrap this entire block in the `try`. You could for instance do the following instead:
```
val report: YarnApplicationReport =
  try {
    getApplicationReport(appId)
  } catch {
    case e: ApplicationNotFoundException =>
      return (..., ...)
  }
val state = report.getYarnApplicationState
```

Comment (Sephiroth-Lin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Done. Thank you!!!

File: yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala [hunk 3]
```diff
@@ -127,23 +127,11 @@ private[spark] class YarnClientSchedulerBackend(
...
+        if (!stopping) {
```
Comment (tgravescs) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: stop() can be called from other areas of the code (like SparkContext.stop()).  Now that the loop isn't checking for it we wouldn't interrupt this thread if that happens and I think we need to handle that case. See the discussions on https://github.com/apache/spark/pull/3143

Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: good point. We might need to put this in a `private var monitorThread` or something and interrupt it in `stop()`. (I would make this method return the thread and set `monitorThread = asyncMonitorApplication()` in `start()`

Comment (Sephiroth-Lin) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: Right. We need to interrupt the thread in stop().

**Comment by andrewor14** [comment]
Nice, LGTM merging into master thanks @Sephiroth-Lin @tgravescs @witgo

**Pull request closed without merging**